File: backend/app/services/futu_mock.py
"""
Mock Futu OpenAPI client.
Mimics futu-api SDK interface so the real client can be swapped in
by replacing this module and setting FUTU_OPEND_HOST in .env.
"""
import random
import uuid
from datetime import datetime

from app.models import (
    AccountBalance,
    Order,
    OrderRequest,
    OrderSide,
    OrderStatus,
    Position,
    RiskCheckResult,
)

# ── Mock data ────────────────────────────────────────────────────

_MOCK_POSITIONS: list[Position] = [
    Position(
        ticker="600519.SH",
        market="CN",
        name="贵州茅台",
        qty=10,
        avg_cost=1680.0,
        current_price=1750.0,
        market_value=17500.0,
        pnl=700.0,
        pnl_pct=4.17,
    ),
    Position(
        ticker="000858.SZ",
        market="CN",
        name="五粮液",
        qty=50,
        avg_cost=140.0,
        current_price=155.0,
        market_value=7750.0,
        pnl=750.0,
        pnl_pct=10.71,
    ),
    Position(
        ticker="AAPL.US",
        market="US",
        name="Apple Inc.",
        qty=20,
        avg_cost=165.0,
        current_price=178.5,
        market_value=3570.0,
        pnl=270.0,
        pnl_pct=8.18,
    ),
]

_MOCK_BALANCE = AccountBalance(
    total_assets=128820.0,
    cash=100000.0,
    market_value=28820.0,
    daily_pnl=1720.0,
    daily_pnl_pct=1.35,
)

_orders: list[Order] = []

# ── Risk control ─────────────────────────────────────────────────

RISK_MAX_POSITION_PCT = 0.10
RISK_MAX_DAILY_LOSS_PCT = 0.02
RISK_MAX_POSITIONS = 5


def check_risk(
    req: OrderRequest,
    balance: AccountBalance,
    positions: list[Position],
) -> RiskCheckResult:
    """Run risk checks. Returns (passed, reason)."""
    order_value = req.qty * req.price

    if order_value / balance.total_assets > RISK_MAX_POSITION_PCT:
        return RiskCheckResult(
            passed=False,
            reason=(
                f"超过单笔仓位限制 {RISK_MAX_POSITION_PCT * 100:.0f}%"
                f"（订单市值 {order_value:,.0f}，总资产 {balance.total_assets:,.0f}）"
            ),
        )

    if balance.daily_pnl < 0:
        loss_pct = abs(balance.daily_pnl) / balance.total_assets
        if loss_pct >= RISK_MAX_DAILY_LOSS_PCT:
            return RiskCheckResult(
                passed=False,
                reason=f"已达日最大亏损限制 {RISK_MAX_DAILY_LOSS_PCT * 100:.0f}%",
            )

    if req.side == OrderSide.BUY:
        existing_tickers = {p.ticker for p in positions}
        if req.ticker not in existing_tickers and len(existing_tickers) >= RISK_MAX_POSITIONS:
            return RiskCheckResult(
                passed=False,
                reason=f"同时持仓数已达上限 {RISK_MAX_POSITIONS} 只",
            )

    return RiskCheckResult(passed=True)


# ── Mock client methods ───────────────────────────────────────────

def get_positions() -> list[Position]:
    return list(_MOCK_POSITIONS)


def get_balance() -> AccountBalance:
    return _MOCK_BALANCE


def place_order(req: OrderRequest) -> Order:
    now = datetime.utcnow().isoformat()
    order = Order(
        order_id=str(uuid.uuid4()),
        ticker=req.ticker,
        market=req.market,
        side=req.side,
        qty=req.qty,
        price=req.price,
        status=OrderStatus.FILLED,
        filled_qty=req.qty,
        filled_price=req.price * random.uniform(0.999, 1.001),
        created_at=now,
        updated_at=now,
        paper_trade=req.paper_trade,
    )
    _orders.append(order)
    return order


def get_orders() -> list[Order]:
    return list(reversed(_orders))

File: server/services/futu-mock.ts
import { randomUUID } from 'crypto'
import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'node:fs'
import { join } from 'node:path'
import type { AccountBalance, Order, OrderRequest, Position, RiskCheckResult } from '../types/index.js'

const DATA_DIR = join(process.cwd(), 'data')
const POSITIONS_FILE = join(DATA_DIR, 'positions.json')
const ORDERS_FILE = join(DATA_DIR, 'orders.json')
const BALANCE_FILE = join(DATA_DIR, 'balance.json')

function ensureDataDir() {
  if (!existsSync(DATA_DIR)) mkdirSync(DATA_DIR, { recursive: true })
}

function loadJSON<T>(file: string, fallback: T): T {
  if (existsSync(file)) {
    try { return JSON.parse(readFileSync(file, 'utf-8')) } catch { /* fall through */ }
  }
  return fallback
}

function saveJSON(file: string, data: unknown) {
  ensureDataDir()
  writeFileSync(file, JSON.stringify(data, null, 2))
}

const DEFAULT_BALANCE: AccountBalance = {
  total_assets: 150000,
  cash: 150000,
  market_value: 0,
  daily_pnl: 0,
  daily_pnl_pct: 0,
}

let _positions: Position[] = loadJSON<Position[]>(POSITIONS_FILE, [])
let _balance: AccountBalance = loadJSON<AccountBalance>(BALANCE_FILE, DEFAULT_BALANCE)
const _ordersArr: Order[] = loadJSON<Order[]>(ORDERS_FILE, [])
const _orders = new Map<string, Order>(_ordersArr.map((o) => [o.order_id, o]))

export function getPositions(): Position[] {
  return _positions
}

export function getBalance(): AccountBalance {
  const market_value = _positions.reduce((s, p) => s + p.market_value, 0)
  _balance.market_value = market_value
  _balance.total_assets = _balance.cash + market_value
  return _balance
}

export function getOrders(): Order[] {
  return Array.from(_orders.values()).sort((a, b) => b.created_at.localeCompare(a.created_at))
}

export function getOrder(orderId: string): Order | undefined {
  return _orders.get(orderId)
}

export function checkRisk(req: OrderRequest): RiskCheckResult {
  if (req.qty <= 0) return { passed: false, reason: '委托数量必须大于 0' }
  if (req.price <= 0) return { passed: false, reason: '委托价格必须大于 0' }
  if (req.side === 'BUY') {
    const orderValue = req.qty * req.price
    if (orderValue > _balance.cash * 0.9)
      return { passed: false, reason: '可用资金不足，委托金额超过可用资金的 90%' }
  }
  if (req.side === 'SELL') {
    const pos = _positions.find((p) => p.ticker === req.ticker)
    if (!pos || pos.qty < req.qty)
      return { passed: false, reason: '持仓数量不足' }
  }
  return { passed: true }
}

function applyFill(order: Order) {
  const { ticker, market, side, qty, price } = order
  const idx = _positions.findIndex((p) => p.ticker === ticker)

  if (side === 'BUY') {
    if (idx >= 0) {
      const p = _positions[idx]
      const newQty = p.qty + qty
      const newAvgCost = (p.qty * p.avg_cost + qty * price) / newQty
      _positions[idx] = {
        ...p,
        qty: newQty,
        avg_cost: newAvgCost,
        current_price: price,
        market_value: newQty * price,
        pnl: (price - newAvgCost) * newQty,
        pnl_pct: ((price - newAvgCost) / newAvgCost) * 100,
      }
    } else {
      _positions.push({
        ticker,
        market,
        name: ticker,
        qty,
        avg_cost: price,
        current_price: price,
        market_value: qty * price,
        pnl: 0,
        pnl_pct: 0,
      })
    }
    _balance.cash -= qty * price
  } else {
    if (idx >= 0) {
      const p = _positions[idx]
      const newQty = p.qty - qty
      if (newQty <= 0) {
        _positions.splice(idx, 1)
      } else {
        _positions[idx] = {
          ...p,
          qty: newQty,
          current_price: price,
          market_value: newQty * price,
          pnl: (price - p.avg_cost) * newQty,
          pnl_pct: ((price - p.avg_cost) / p.avg_cost) * 100,
        }
      }
      _balance.cash += qty * price
    }
  }

  saveJSON(POSITIONS_FILE, _positions)
  saveJSON(BALANCE_FILE, _balance)
}

export function placeOrder(req: OrderRequest): Order {
  const orderId = randomUUID()
  const now = new Date().toISOString()
  const order: Order = {
    order_id: orderId,
    ticker: req.ticker,
    market: req.market,
    side: req.side,
    qty: req.qty,
    price: req.price,
    status: 'PENDING',
    filled_qty: 0,
    filled_price: 0,
    created_at: now,
    updated_at: now,
    paper_trade: req.paper_trade ?? true,
  }
  _orders.set(orderId, order)
  saveJSON(ORDERS_FILE, Array.from(_orders.values()))

  Promise.resolve().then(() => {
    const filled: Order = {
      ...order,
      status: 'FILLED',
      filled_qty: req.qty,
      filled_price: req.price,
      updated_at: new Date().toISOString(),
    }
    _orders.set(orderId, filled)
    saveJSON(ORDERS_FILE, Array.from(_orders.values()))
    applyFill(filled)
  })

  return order
}
